docs/rfc/0001-exporter-interface.md

```diff
@@ -0,0 +1,194 @@
+that will be used for each parameter. A toml file stored inside the Indexer data directory stores all of the config data
+for a given plugin. The Indexer will use the type specified in the Indexer config to look for a plugin config which
+satisfies that class, and then load that as the selected plugin. Supplying multiple plugin configs for the selected type
+of exporter will result in undefined behavior--a random config will be chosen.
+```
+[Plugin.Exporter]
+name = "postgresql-exporter"
+type = "IndexerPostgresqlExporter"
+username = "foo"
+password = "bar"
+host = "127.0.0.1"
+port = "1234"
+dbname = "indexer"
+```
+
+**Indexer Config**
+
+The Indexer's config will need to be changed to incorporate plugins. Future RFCs can decide how to string multiple
+pipelines and plugins together in interesting ways. For now, we will have a single pipeline per indexer--i.e. running the
+indexer will only run a single data pipeline and therefore have a single exporter. In that way the initial changes will
+be easy to make--one new field in the config which provides an internal plugin name. Config files for the selected
+plugin will then be resolved and parsed at runtime. The default configuration for Indexer will select the Postgresql
+plugin in order to maintain backwards compatibility with the existing Indexer pipeline.
+
+***Plugin Execution Framework***
+The plugin execution framework will have the following features.
+* Guaranteed block processing
+
+The existing Indexer implementation uses the postgres database for multiple purposes--both for storing data about the
+state of the system such as the most recently processed round or the migration status, as well as for storing the actual
+block data being exported. The new plugin execution framework will not be able to rely on the export destination to
+store state, so it will use its own data store to flush state to disk. One of the main purposes of this will be to
+ensure only-once, in order block processing, but it can also be exposed to plugins that need to store state, or for
+other purposes as they are uncovered.
+
+* Built In Data Format Definitions
+
+Though not being defined here, we can imagine that intermediate plugins in our execution framework might perform common
+data operations such as aggregation, filtering, annotation, etc. The result of some of these operations will be a subset
+of full block data, while others may result in block data which does not conform to the block specification at all.
+In order to accommodate customization of the data, we will provide multiple data formats--the standard Block interface
+used today, as well as more generic forms that plugins can deserialize based on their needs. A given plugin will need to
+specify the data format it expects for both input data and output if it provides any (exporters obviously don't have
+output data in this sense). The system will ensure that the constructed pipeline has compatible data formats during
+initialization.
```

Comment:
I think it would be good to define an initial format, and extend it in the future.

Today, we have "Block" and "StateDelta". We're missing "Certificate", but we could add that in as well. I think they would all be nullable.

Reply:
Noted. I will add an example here.